File: app/src/main/java/com/erkaslan/moviedb/ui/HomeFragment.java
package com.erkaslan.moviedb.ui;

import android.content.ClipData;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.erkaslan.moviedb.MovieApi;
import com.erkaslan.moviedb.R;
import com.erkaslan.moviedb.data.Movie;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;


public class HomeFragment extends Fragment{

    private static final String TAG = "Home";

    public Button buttonSearch;
    public TextView textTitle, textYear, textNotFound;
    public EditText inputMovie;
    public ProgressBar progressBar;

    private RecyclerView movie;
    private RecyclerView.Adapter movieAdapter;
    private RecyclerView.LayoutManager movieLayoutManager;

    ArrayList<Movie> movieList;


    public HomeFragment() {
        super(R.layout.fragment_home);
    }

    public static HomeFragment newInstance() {
        HomeFragment fragment = new HomeFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View v = inflater.inflate(R.layout.fragment_home, container, false);
        buttonSearch = (Button) v.findViewById(R.id.buttonSearch);
        textTitle = (TextView) v.findViewById(R.id.textTitle);
        textYear = (TextView) v.findViewById(R.id.textYear);
        textNotFound = (TextView) v.findViewById(R.id.textNotFound);
        inputMovie = (EditText) v.findViewById(R.id.inputMovie);
        progressBar = (ProgressBar) v.findViewById(R.id.progressBar);
        movie = (RecyclerView) v.findViewById(R.id.recylerMovie);

        initializeRecyclerView();
        SharedViewModel sharedViewModel = new SharedViewModel(this.getActivity().getApplication());

        //Search button click sets movie set
        buttonSearch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                progressBar.setVisibility(View.VISIBLE);
                Log.v(TAG, "CLICKED");
                if(inputMovie.getText().toString().isEmpty()) {
                    movie.setVisibility(View.GONE);
                    progressBar.setVisibility(View.GONE);
                    textNotFound.setText("Type in something, millions of fascinating movies are out there");
                    textNotFound.setVisibility(View.VISIBLE);
                }else{
                    sharedViewModel.setMovieList(inputMovie.getText().toString());
                    textNotFound.setVisibility(View.INVISIBLE);
                    movie.setVisibility(View.VISIBLE);
                }
            }
        });


        //observer sets recylerview to movie set once movelist is changed
        sharedViewModel.getMovieList().observe(getViewLifecycleOwner(), new Observer<ArrayList<Movie>>() {
            @Override
            public void onChanged(ArrayList<Movie> moviesAll) {

                progressBar.setVisibility(View.GONE);

                if (moviesAll.size() == 0) {
                    Log.v(TAG, "****movie not found");
                    movie.setVisibility(View.GONE);
                    textNotFound.setText("No movie found...are you sure about the title?");
                    textNotFound.setVisibility(View.VISIBLE);
                }else{
                    Log.v(TAG, "onChanged: XXX");
                    Log.v(TAG, "movies number: " + String.valueOf(moviesAll.size()));
                    textNotFound.setVisibility(View.INVISIBLE);

                    movieAdapter = new MovieListAdapter(getContext(), moviesAll);
                    movie.setAdapter(movieAdapter);
                    movie.setVisibility(View.VISIBLE);
                    movieAdapter.notifyDataSetChanged();
                }
            }
        });

        // Inflate the layout for this fragment
        return v;
    }

    //Sets up the movie, adapter etc.
    private void initializeRecyclerView() {
        movieList = new ArrayList<>( );
        movie.setNestedScrollingEnabled(false);
        movie.setHasFixedSize(false);
        movieLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        movie.setLayoutManager(movieLayoutManager);
        movieAdapter = new MovieListAdapter(getContext(), movieList);
        movie.setAdapter(movieAdapter);
        Log.v(TAG, "****adapter initialized");
    }

}

File: app/src/main/java/com/erkaslan/moviedb/ui/HomeFragment.kt
package com.erkaslan.moviedb.ui

import com.erkaslan.moviedb.R
import androidx.recyclerview.widget.RecyclerView
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import com.erkaslan.moviedb.data.Movie
import com.erkaslan.moviedb.databinding.FragmentHomeBinding
import java.util.ArrayList

class HomeFragment : Fragment() {

    private var movieAdapter: RecyclerView.Adapter<*>? = null
    private var movieLayoutManager: RecyclerView.LayoutManager? = null
    var movieList: ArrayList<Movie>? = null
    private var _binding: FragmentHomeBinding? = null
    val binding get() = _binding!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeRecyclerView(binding.recylerMovie)
        val sharedViewModel = SharedViewModel(this.requireActivity().application)

        if (binding.inputMovie.text.toString().isEmpty()) {
            binding.recylerMovie.visibility = View.GONE
            binding.progressBar.visibility = View.GONE
        } else {
            sharedViewModel.setMovieList(binding.inputMovie.text.toString())
            binding.textNotFound.visibility = View.INVISIBLE
            binding.recylerMovie.visibility = View.VISIBLE
        }

        //Search button click sets movie set
        binding.buttonSearch.setOnClickListener {
            binding.progressBar.visibility = View.VISIBLE
            if (binding.inputMovie.text.toString().isEmpty()) {
                binding.recylerMovie.visibility = View.GONE
                binding.progressBar.visibility = View.GONE
                binding.textNotFound.text =
                    "Type in something, millions of fascinating movies are out there"
                binding.textNotFound.visibility = View.VISIBLE
            } else {
                sharedViewModel.setMovieList(binding.inputMovie.text.toString())
                binding.textNotFound.visibility = View.INVISIBLE
                binding.recylerMovie.visibility = View.VISIBLE
            }
        }

        //observer sets recylerview to movie set once movelist is changed
        sharedViewModel.moviesAll.observe(viewLifecycleOwner, { moviesAll ->
            binding.progressBar.visibility = View.GONE
            if (moviesAll?.size == 0) {
                binding.recylerMovie.visibility = View.GONE
                binding.textNotFound.text = "No movie found...are you sure about the title?"
                binding.textNotFound.visibility = View.VISIBLE
            } else {
                binding.textNotFound.visibility = View.INVISIBLE
                movieAdapter = context?.let {
                    MovieListAdapter(it, moviesAll ?: arrayListOf<Movie>(), object : MovieListAdapter.ItemClickListener {
                        override fun itemClick(position: Int, movie: Movie) {
                            val bundle = Bundle()
                            bundle.putString("title", movie.title)
                            bundle.putString("imdbID", movie.imdbId)
                            bundle.putString("poster", movie.poster)
                            bundle.putString("year", movie.year)
                            Navigation.findNavController(binding.root)
                                .navigate(R.id.action_homeFragment_to_detailFragment, bundle)
                        }
                    })
                }
                binding.recylerMovie.adapter = movieAdapter
                binding.recylerMovie.visibility = View.VISIBLE
                movieAdapter?.let { it.notifyDataSetChanged() }
            }
        })
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        binding.lifecycleOwner = activity
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    //Sets up the movie, adapter etc.
    private fun initializeRecyclerView(movie: RecyclerView) {
        movieList = ArrayList()
        movie.isNestedScrollingEnabled = false
        movie.setHasFixedSize(false)
        movieLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        movie.layoutManager = movieLayoutManager
        movieAdapter = context?.let {
            MovieListAdapter(it, movieList!!, object : MovieListAdapter.ItemClickListener {
                override fun itemClick(position: Int, movie: Movie) {

                }
            })
        }
        movie.adapter = movieAdapter
    }

    companion object {
        private const val TAG = "Home"
        fun newInstance(): HomeFragment {
            val fragment = HomeFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}